gadgets/deadlock/program.bpf.c

```diff
@@ -9,68 +9,209 @@
 #include <gadget/macros.h>
 #include <gadget/mntns_filter.h>
 
-enum operation { lock, unlock };
+#define MAX_HELD_MUTEXES 16
+#define MAX_ENTRIES 65536
+#define MAX_STACK_DEPTH 50
+
+/* Stack map */
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__type(key, u32);
+	__uint(max_entries, 256);
+	__uint(value_size, MAX_STACK_DEPTH * sizeof(u64));
+} stackmap SEC(".maps");
+
+// Represents a mutex held by a thread.
+struct held_mutex {
+	u64 mutex;
+	u64 stack_id;
+};
+
+/* Map of thread ID to an array of held_mutex structs */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value,
+	       struct held_mutex[MAX_HELD_MUTEXES]); // array of held mutexes
+} thread_to_held_mutexes SEC(".maps");
+
+// Key type for edges. Represents an edge from mutex1 to mutex2.
+struct edges_key {
+	__u64 mutex1;
+	__u64 mutex2;
```

Comment:
I am wondering if the edge_key should include the process id (or some other identifier of the struct mm) to avoid false positive when two completely separate processes happen to use mutexes with the same address.